odxtools/diaglayer.py

```diff
@@ -843,67 +845,90 @@ def _build_coded_prefix_tree(self):
+            # possible responses. We need to consider the global
+            # negative responses here, because they might contain
+            # MATCHING-REQUEST parameters. If these global responses
+            # do not contain such parameters, this will potentially
+            # result in an enormous amount of decoded messages for
+            # global negative responses. (I.e., one for each
+            # service. This can be avoided by specifying the
+            # corresponding request for `decode_response()`.)
             request_prefix = s.request.coded_const_prefix()
-            prefixes = [request_prefix] + [
-                message.coded_const_prefix(request_prefix=request_prefix)
-                for message in chain(s.positive_responses, s.negative_responses)
+            prefixes = [request_prefix]
+            prefixes += [
+                x.coded_const_prefix(request_prefix=request_prefix) for x in chain(
+                    s.positive_responses, s.negative_responses, self.global_negative_responses)
             ]
             for coded_prefix in prefixes:
-                # Traverse prefix tree
-                sub_tree = prefix_tree
-                for b in coded_prefix:
-                    if sub_tree.get(b) is None:
-                        sub_tree[b] = {}
-                    sub_tree = sub_tree.get(b)
-
-                    assert isinstance(
-                        sub_tree,
-                        dict), f"{sub_tree} has type {type(sub_tree)}. How did this happen?"
-                # Add service as leaf to prefix tree
-                if sub_tree.get(-1) is None:
-                    sub_tree[-1] = [s]
-                else:
-                    sub_tree[-1].append(s)
+                self._extend_prefix_tree(prefix_tree, coded_prefix, s)
+
         return prefix_tree
 
-    def _find_services_for_uds(self, message: Union[bytes, bytearray]):
+    @staticmethod
+    def _extend_prefix_tree(prefix_tree: PrefixTree, coded_prefix: bytes,
+                            service: DiagService) -> None:
+
+        # make sure that tree has an entry for the given prefix
+        sub_tree = prefix_tree
+        for b in coded_prefix:
+            if b not in sub_tree:
+                sub_tree[b] = {}
+            sub_tree = cast(PrefixTree, sub_tree[b])
+
+        # Store the object as in the prefix tree. This is done by
+        # assigning the list of possible objects to the key -1 of the
+        # dictionary (this is quite hacky...)
+        if sub_tree.get(-1) is None:
+            sub_tree[-1] = [service]
+        else:
+            cast(List[DiagService], sub_tree[-1]).append(service)
+
+    def _find_services_for_uds(self, message: Union[bytes, bytearray]) -> List[DiagService]:
         if not hasattr(self, "_prefix_tree"):
             # Compute the prefix tree the first time this decode function is called.
             self._prefix_tree = self._build_coded_prefix_tree()
```

Comment:
maybe use the cached decorator? it usual simplify this kind of logic

Reply:
thanks. implemented: 495cdf83ed657c3b0f3c0a166637351cae6b77de